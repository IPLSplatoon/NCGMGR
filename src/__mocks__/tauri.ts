export const mockTauriEvent = {
    listen: jest.fn()
}

jest.mock('@tauri-apps/api/event', () => mockTauriEvent)

export const mockTauri = {
    invoke: jest.fn()
}

jest.mock('@tauri-apps/api/tauri', () => mockTauri)

export const mockTauriDialog = {
    open: jest.fn()
}

jest.mock('@tauri-apps/api/dialog', () => mockTauriDialog)

export const mockTauriFs = {
    readDir: jest.fn(),
    readTextFile: jest.fn(),
    removeDir: jest.fn(),
    removeFile: jest.fn(),
    createDir: jest.fn(),
    writeFile: jest.fn()
}

jest.mock('@tauri-apps/api/fs', () => mockTauriFs)

export const mockTauriOs = {
    type: jest.fn()
}

jest.mock('@tauri-apps/api/os', () => mockTauriOs)

export const mockTauriShell = {
    open: jest.fn()
}

jest.mock('@tauri-apps/api/shell', () => mockTauriShell)
