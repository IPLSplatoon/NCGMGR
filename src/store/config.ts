import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const configStore = createStore<Configuration>({
    state: {
        installPath: ''
    },
    mutations: {
        setInstallPath: (state, installPath: string) => {
            state.installPath = installPath
        },
        setState: (state, config: Configuration) => {
            Object.assign(state, config)
        }
    },
    actions: {
        load: store => {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            store.commit('setState', {
                ...store.state,
                ...config
            })
        },
        persist: store => {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(store.state))
        }
    }
})

export const configStoreKey: InjectionKey<Store<Configuration>> = Symbol()

export function useConfigStore (): Store<Configuration> {
    return useStore(configStoreKey)
}
