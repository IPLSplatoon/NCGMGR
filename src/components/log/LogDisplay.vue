<template>
    <div class="log-display">
        <span
            v-for="(line, i) in log"
            :key="`log-line_${i}`"
            :class="`type-${line.type ?? 'unknown'}`"
        >
            <template
                v-for="(part, j) in line.messageParts"
            >
                <template v-if="!part.was_processed">
                    {{ part.content }}
                </template>
                <span
                    v-else
                    :key="`log-line_${i}_${j}`"
                    :class="classListFromLogLine(part)"
                >
                    {{ part.content }}
                </span>
            </template>
        </span>
    </div>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, onUnmounted, watch } from 'vue'
import { useLogStore } from '@/store/logStore'
import Anser from 'anser'

export default defineComponent({
    name: 'LogDisplay',

    props: {
        logKey: {
            type: String,
            required: true
        }
    },

    setup (props) {
        const logStore = useLogStore()

        watch(() => props.logKey, (newValue, oldValue) => {
            if (oldValue) {
                logStore.unlisten(oldValue)
            }
            logStore.listen(newValue)
        })

        onMounted(() => {
            logStore.listen(props.logKey)
        })

        onUnmounted(() => {
            logStore.unlisten(props.logKey)
        })

        return {
            log: computed(() => {
                const log = logStore.lines[props.logKey] ?? []
                return log.map(line => ({
                    ...line,
                    messageParts: Anser.ansiToJson(line.message, { use_classes: true })
                }))
            }),
            classListFromLogLine(line: Anser.AnserJsonEntry) {
                const result = []

                if (line.bg != null) {
                    result.push(`${line.bg}-bg`)
                }
                if (line.fg != null) {
                    result.push(`${line.fg}-fg`)
                }

                return result
            }
        }
    }
})
</script>

<style lang="scss" scoped>
@import '../../styles/layout';
@import '../../styles/text';
@import '../../styles/colors';

.log-display {
    font-family: monospace;
    width: calc(100% - 8px);
    max-height: 350px;
    text-align: left;
    border: 1px solid $input-color;
    padding: 4px 2px;
    min-height: 64px;
    overflow: scroll;
    color: var(--ipl-body-text-color);
    background-color: var(--ipl-bg-primary);

    > span {
        margin-left: 6px;
        display: block;
        white-space: nowrap;

        &.type-error {
            color: $error-color;
        }

        .foo {
            color: red;
        }
    }
}
</style>
