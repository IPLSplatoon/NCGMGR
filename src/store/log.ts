import { LogEvent } from '@/types/log'
import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'
import { Event, listen, UnlistenFn } from '@tauri-apps/api/event'

export interface LogStore {
    lines: Record<string, LogEvent[]>
    completed: Record<string, boolean>
}

const unlistenFns: Record<string, UnlistenFn> = {}

export const logStore = createStore<LogStore>({
    state: {
        lines: {},
        completed: {}
    },
    mutations: {
        insertLine (state, { line, key }: { line: LogEvent, key: string }) {
            if (!state.lines[key]) {
                state.lines[key] = []
            }
            state.lines[key].push(line)
        },
        reset (state, key: string) {
            state.lines[key] = []
            state.completed[key] = false
        },
        setCompleted (state, { key, completed }: { key: string, completed: boolean }) {
            state.completed[key] = completed
        }
    },
    actions: {
        listen (store, key: string) {
            if (unlistenFns[key]) {
                return
            }
            return listen(`log:${key}`, (event: Event<LogEvent>) => {
                store.commit('insertLine', { line: event.payload, key })
            }).then(unlisten => {
                unlistenFns[key] = unlisten
            })
        },
        unlisten (store, key: string) {
            if (unlistenFns[key]) {
                unlistenFns[key]()
            }
        },
        logPromiseResult (store, { promise, key, noLogOnSuccess }: { promise: Promise<unknown>, key: string, noLogOnSuccess: boolean }) {
            promise.then(() => {
                if (!noLogOnSuccess) {
                    store.commit('insertLine', {
                        line: {
                            message: 'Success!',
                            type: 'success'
                        },
                        key
                    })
                }
            }).catch(e => {
                store.commit('insertLine', {
                    line: {
                        message: String(e),
                        type: 'error'
                    },
                    key
                })
            }).finally(() => {
                store.commit('setCompleted', { key, completed: true })
            })
        }
    }
})

export const logStoreKey: InjectionKey<Store<LogStore>> = Symbol()

export function useLogStore (): Store<LogStore> {
    return useStore(logStoreKey)
}
