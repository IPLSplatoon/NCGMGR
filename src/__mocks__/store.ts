import { createStore, Store } from 'vuex'
import { Configuration } from '@/store/config'
import { LogStore } from '@/store/log'
import { NodecgStatus, NodecgStore } from '@/store/nodecg'

export function createConfigStore (): Store<Configuration> {
    return createStore<Configuration>({
        state: {
            installPath: '/install/path'
        },
        actions: {
            load: jest.fn(),
            persist: jest.fn()
        },
        mutations: {
            setInstallPath: jest.fn()
        }
    })
}

export function createLogStore (): Store<LogStore> {
    return createStore<LogStore>({
        state: {
            lines: [],
            completed: false
        },
        mutations: {
            reset: jest.fn(),
            insertLine: jest.fn(),
            setCompleted: jest.fn()
        },
        actions: {
            logPromiseResult: jest.fn()
        }
    })
}

export function createNodecgStore (): Store<NodecgStore> {
    return createStore<NodecgStore>({
        state: {
            status: {
                status: NodecgStatus.UNKNOWN,
                message: '',
                bundlesLoading: false
            },
            bundles: []
        },
        actions: {
            checkNodecgStatus: jest.fn(),
            getBundleList: jest.fn()
        }
    })
}
