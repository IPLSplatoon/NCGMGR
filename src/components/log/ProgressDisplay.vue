<template>
    <ipl-progress-bar
        v-if="progress != null"
        :value="progress"
        :color="progressBarColor"
        background-color="primary"
    />
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { useLogStore } from '@/store/logStore'
import { computed } from 'vue'
import { IplProgressBar } from '@iplsplatoon/vue-components'
import { ActionState } from '@/types/log'
import { themeColors } from '@/styles/colors'

export default defineComponent({
    name: 'ProgressDisplay',

    components: { IplProgressBar },

    props: {
        logKey: {
            type: String,
            required: true
        }
    },

    setup (props) {
        const logStore = useLogStore()
        const actionState = computed(() => logStore.actionStates[props.logKey])

        return {
            progress: computed(() => {
                const entry = logStore.progressEntries[props.logKey]
                if (entry?.maxStep == null) {
                    return null
                }

                if (actionState.value != null && actionState.value !== ActionState.INCOMPLETE) {
                    return 100
                }

                return !entry ? 0 : ((entry.step ?? 0) / entry.maxStep) * 100
            }),
            progressBarColor: computed(() => {
                switch (actionState.value) {
                    case ActionState.COMPLETED_ERROR:
                        return themeColors.red
                    case ActionState.COMPLETED_SUCCESS:
                        return themeColors.green
                    default:
                        return themeColors.blue
                }
            })
        }
    }
})
</script>
