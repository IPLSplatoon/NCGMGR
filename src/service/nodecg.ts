import { readDir, readTextFile, removeDir, removeFile } from '@tauri-apps/api/fs'
import { PackageSchema } from '@/types/package'
import isEmpty from 'lodash/isEmpty'
import { InstallStatus } from '@/store/nodecg'
import { invoke } from '@tauri-apps/api/tauri'

export async function getNodecgStatus (directory: string): Promise<{ status: InstallStatus, message: string }> {
    if (isEmpty(directory?.trim())) {
        return {
            status: InstallStatus.UNABLE_TO_INSTALL,
            message: 'Please select an installation directory.'
        }
    }

    const dir = await readDir(directory)
    if (dir.length < 1) {
        return {
            status: InstallStatus.READY_TO_INSTALL,
            message: 'Directory is empty. Ready to install...'
        }
    } else {
        const packageFile = dir.find(entry => entry.name === 'package.json')
        if (packageFile) {
            const packageJson: PackageSchema = JSON.parse(await readTextFile(packageFile.path))
            if (packageJson.name === 'nodecg') {
                return {
                    status: InstallStatus.INSTALLED,
                    message: `Found NodeCG v${packageJson.version}.`
                }
            } else {
                return {
                    status: InstallStatus.UNABLE_TO_INSTALL,
                    message: `Found unknown package "${packageJson.name}".`
                }
            }
        } else {
            return {
                status: InstallStatus.UNABLE_TO_INSTALL,
                message: 'Could not find package.json.'
            }
        }
    }
}

export interface Bundle {
    name: string
    version: string
}

export async function getBundles (directory: string): Promise<Bundle[]> {
    if (isEmpty(directory.trim())) {
        throw new Error('No bundle directory provided.')
    }

    const bundlesDir = await readDir(directory + '/bundles', { recursive: true })
    return Promise.all(
        bundlesDir
            .filter(entry => entry.children?.some(child => child.name === 'package.json'))
            .map(async (dir) => {
                const packageFile = dir.children?.find(child => child.name === 'package.json')?.path
                if (!packageFile) {
                    throw new Error(`Missing package.json in directory ${dir.name}`)
                }

                const packageJson = JSON.parse(await readTextFile(packageFile)) as PackageSchema

                return {
                    name: packageJson.name,
                    version: packageJson.version
                }
            }))
}

export async function getBundleVersions (bundleName: string, nodecgPath: string): Promise<string[]> {
    return invoke('fetch_bundle_versions', { bundleName, nodecgPath })
}

export async function configFileExists (bundleName: string, nodecgPath: string): Promise<boolean> {
    const configDir = await readDir(`${nodecgPath}/cfg`)
    return configDir.some(item => item.name === `${bundleName}.json`)
}

export async function removeBundle (bundleName: string, nodecgPath: string): Promise<[void, void]> {
    return Promise.all([
        removeDir(`${nodecgPath}/bundles/${bundleName}`, { recursive: true }),
        (async () => {
            if (await configFileExists(bundleName, nodecgPath)) {
                return removeFile(`${nodecgPath}/cfg/${bundleName}.json`)
            }
        })()
    ])
}
