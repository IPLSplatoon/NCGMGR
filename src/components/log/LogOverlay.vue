<template>
    <mgr-overlay
        v-model:visible="modelVisible"
        class="log-overlay"
        :no-background-close="!completed"
        :max-width="moreDetailsShown"
    >
        <div>
            <ipl-dialog-title
                :title="title"
                class="m-b-8"
                :closing-disabled="!completed"
                @close="modelVisible = false"
            />
            <ipl-space>
                <div class="m-b-8">
                    {{ progressMessage }}
                </div>
                <progress-display :log-key="logKey" />
            </ipl-space>
            <ipl-expanding-space
                v-model:expanded="moreDetailsShown"
                title="More details"
                class="m-t-8 max-width"
            >
                <log-display
                    :log-key="logKey"
                    class="m-t-8"
                />
            </ipl-expanding-space>
        </div>
    </mgr-overlay>
</template>

<script lang="ts">
import { computed, defineComponent, ref } from 'vue'
import MgrOverlay from '@/components/mgr/MgrOverlay.vue'
import { IplDialogTitle, IplExpandingSpace, IplSpace } from '@iplsplatoon/vue-components'
import { useLogStore } from '@/store/logStore'
import LogDisplay from '@/components/log/LogDisplay.vue'
import ProgressDisplay from '@/components/log/ProgressDisplay.vue'
import { ActionState } from '@/types/log'

export default defineComponent({
    name: 'LogOverlay',

    components: { IplSpace, IplDialogTitle, IplExpandingSpace, ProgressDisplay, LogDisplay, MgrOverlay },

    props: {
        title: {
            type: String,
            required: true
        },
        visible: {
            type: Boolean,
            required: true
        },
        logKey: {
            type: String,
            required: true
        }
    },

    emits: ['update:visible'],

    setup (props, { emit }) {
        const logStore = useLogStore()

        const modelVisible = computed({
            get () {
                return props.visible
            },
            set (newValue: boolean) {
                emit('update:visible', newValue)
            }
        })
        const completed = computed(() => logStore.actionStates[props.logKey] !== ActionState.INCOMPLETE)

        return {
            progressMessage: computed(() => {
                switch (logStore.actionStates[props.logKey]) {
                    case ActionState.COMPLETED_ERROR:
                        return 'An error has occurred.'
                    case ActionState.COMPLETED_SUCCESS:
                        return 'Done!'
                    default:
                        return 'Please wait...'
                }
            }),
            moreDetailsShown: ref(false),
            completed,
            modelVisible
        }
    }
})
</script>
