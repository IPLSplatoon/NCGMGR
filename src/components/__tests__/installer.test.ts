import { mockTauri, mockTauriDialog } from '@/__mocks__/tauri'
import { config, flushPromises, shallowMount } from '@vue/test-utils'
import { configStoreKey } from '@/store/config'
import { createConfigStore, createLogStore, createNodecgStore } from '@/__mocks__/store'
import { logStoreKey } from '@/store/log'
import Installer from '../Installer.vue'
import { NodecgStatus, nodecgStoreKey } from '@/store/nodecg'

describe('Installer', () => {
    config.global.stubs = {
        IplSpace: false
    }

    it('matches snapshot', () => {
        const wrapper = shallowMount(Installer, {
            global: {
                plugins: [
                    [createConfigStore(), configStoreKey],
                    [createNodecgStore(), nodecgStoreKey]
                ]
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('handles directory selection', async () => {
        const store = createConfigStore()
        jest.spyOn(store, 'dispatch')
        jest.spyOn(store, 'commit')
        const wrapper = shallowMount(Installer, {
            global: {
                plugins: [
                    [store, configStoreKey],
                    [createNodecgStore(), nodecgStoreKey]
                ]
            }
        })
        mockTauriDialog.open.mockResolvedValue('/new/path')

        wrapper.getComponent('[data-test="install-directory-select-button"]').vm.$emit('click')
        await flushPromises()

        expect(mockTauriDialog.open).toHaveBeenCalledWith({ directory: true })
        expect(store.commit).toHaveBeenCalledWith('setInstallPath', '/new/path')
        expect(store.dispatch).toHaveBeenCalledWith('persist')
        expect(store.dispatch).toHaveBeenCalledTimes(1)
    })

    it('handles directory selection getting cancelled', async () => {
        const store = createConfigStore()
        jest.spyOn(store, 'dispatch')
        jest.spyOn(store, 'commit')
        const wrapper = shallowMount(Installer, {
            global: {
                plugins: [
                    [store, configStoreKey],
                    [createNodecgStore(), nodecgStoreKey]
                ]
            }
        })
        mockTauriDialog.open.mockResolvedValue(null)

        wrapper.getComponent('[data-test="install-directory-select-button"]').vm.$emit('click')
        await flushPromises()

        expect(store.commit).not.toHaveBeenCalled()
        expect(store.dispatch).not.toHaveBeenCalled()
    })

    it('disables installation if installation is not possible', async () => {
        const nodecgStore = createNodecgStore()
        nodecgStore.state.status.status = NodecgStatus.UNABLE_TO_INSTALL
        const store = createConfigStore()
        const wrapper = await shallowMount(Installer, {
            global: {
                plugins: [
                    [store, configStoreKey],
                    [nodecgStore, nodecgStoreKey]
                ]
            }
        })
        await flushPromises()

        expect(wrapper.getComponent('[data-test="install-button"]').attributes().disabled).toEqual('true')
    })

    it('enables installation if installation is possible', async () => {
        const nodecgStore = createNodecgStore()
        nodecgStore.state.status.status = NodecgStatus.READY_TO_INSTALL
        const store = createConfigStore()
        const wrapper = await shallowMount(Installer, {
            global: {
                plugins: [
                    [store, configStoreKey],
                    [nodecgStore, nodecgStoreKey]
                ]
            }
        })
        await flushPromises()

        expect(wrapper.getComponent('[data-test="install-button"]').attributes().disabled).toEqual('false')
    })

    it('disables installation if installation is completed', async () => {
        const nodecgStore = createNodecgStore()
        nodecgStore.state.status.status = NodecgStatus.INSTALLED
        const store = createConfigStore()
        const wrapper = await shallowMount(Installer, {
            global: {
                plugins: [
                    [store, configStoreKey],
                    [nodecgStore, nodecgStoreKey]
                ]
            }
        })
        await flushPromises()

        expect(wrapper.getComponent('[data-test="install-button"]').attributes().disabled).toEqual('true')
    })

    it('disables installation if installation status is unknown', async () => {
        const nodecgStore = createNodecgStore()
        nodecgStore.state.status.status = NodecgStatus.UNKNOWN
        const store = createConfigStore()
        const wrapper = await shallowMount(Installer, {
            global: {
                plugins: [
                    [store, configStoreKey],
                    [nodecgStore, nodecgStoreKey]
                ]
            }
        })
        await flushPromises()

        expect(wrapper.getComponent('[data-test="install-button"]').attributes().disabled).toEqual('true')
    })

    it('handles installation', () => {
        const configStore = createConfigStore()
        configStore.state.installPath = '/install/path'
        const logStore = createLogStore()
        jest.spyOn(logStore, 'commit')
        mockTauri.invoke.mockResolvedValue({})
        const wrapper = shallowMount(Installer, {
            global: {
                plugins: [
                    [configStore, configStoreKey],
                    [logStore, logStoreKey],
                    [createNodecgStore(), nodecgStoreKey]
                ]
            }
        })

        wrapper.getComponent('[data-test="install-button"]').vm.$emit('click')

        expect(logStore.commit).toHaveBeenCalledWith('reset')
        expect(mockTauri.invoke).toHaveBeenCalledWith('install_nodecg', { path: '/install/path' })
    })
})
