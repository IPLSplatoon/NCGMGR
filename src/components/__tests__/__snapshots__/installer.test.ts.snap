// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Installer matches snapshot 1`] = `
<div class="ipl-space color-dark layout vertical">
  <div>Installation folder: /install/path</div>
  <status-row-stub color="gray" class="m-t-8"></status-row-stub>
  <div class="layout horizontal m-t-8">
    <ipl-button-stub label="Select folder" color="blue" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" nobackground="false" tooltip="" data-test="install-directory-select-button"></ipl-button-stub>
    <ipl-button-stub label="Install" color="green" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" nobackground="false" tooltip="" data-test="install-button" class="m-l-8"></ipl-button-stub>
  </div>
</div>
<log-overlay-stub title="Installing..." visible="false"></log-overlay-stub>
`;
